### PR TITLE
Implement Python Import Caching

### PR DESCRIPTION
**Pull request by mxwli**
```
py::module::import("module").attr("attribute");
```
becomes
```
kuzu::pycache::import_cache->module.attribute();
```

Now that modules are shared, it's required to acquire the Global Interpreter Lock before accessing modules.

Technically, it's possible to skip acquiring this lock if we know the module is already cached, but that isn't implemented here.

### CODE REVIEW

File: tools/python_api/src_cpp/pandas/pandas_scan.cpp [hunk 1]
```diff
@@ -127,10 +129,10 @@ std::unique_ptr<Value> tryReplacePD(py::dict& dict, py::str& tableName) {
 }
 
 std::unique_ptr<common::Value> replacePD(common::Value* value) {
-    py::gil_scoped_acquire acquire;
     auto pyTableName = py::str(value->getValue<std::string>());
     // Here we do an exhaustive search on the frame lineage.
-    auto currentFrame = py::module::import("inspect").attr("currentframe")();
+    py::gil_scoped_acquire acquire;
+    auto currentFrame = import_cache->inspect.currentframe()();
```
Comment (acquamarin) [inline tools/python_api/src_cpp/pandas/pandas_scan.cpp:135]: You can release the gil right after accesssing the cache?

File: tools/python_api/src_cpp/py_connection.cpp [hunk 2]
```diff
@@ -175,10 +176,10 @@ std::unordered_map<std::string, std::unique_ptr<Value>> transformPythonParameter
 }
 
 Value transformPythonValue(py::handle val) {
-    auto datetime_mod = py::module::import("datetime");
-    auto datetime_datetime = datetime_mod.attr("datetime");
-    auto time_delta = datetime_mod.attr("timedelta");
-    auto datetime_date = datetime_mod.attr("date");
+    py::gil_scoped_acquire acquire;
```
Comment (acquamarin) [inline tools/python_api/src_cpp/py_connection.cpp:179]: I don't think you have to hold a gil there.
You only have to hold the gil when the cache is not loaded?

File: tools/python_api/src_cpp/py_conversion.cpp [hunk 3]
```diff
@@ -1,18 +1,19 @@
...
+using namespace kuzu::pycache;
 
 PythonObjectType getPythonObjectType(py::handle& ele) {
-    py::object pandas = py::module::import("pandas");
-    auto pandasNa = pandas.attr("NA");
-    auto pandasNat = pandas.attr("NaT");
-    py::object datetime = py::module::import("datetime");
-    auto pyDateTime = datetime.attr("datetime");
-    auto pyDate = datetime.attr("date");
+    py::gil_scoped_acquire acquire;
```
Comment (acquamarin) [inline tools/python_api/src_cpp/py_conversion.cpp:12]: ditto, i don't see duckdb holds the gil there

File: tools/python_api/src_cpp/include/cached_import/py_cached_import.h [hunk 4]
```diff
@@ -0,0 +1,32 @@
+#pragma once
+
+#include <vector>
+
+#include "py_cached_modules.h"
+
+namespace kuzu {
+namespace pycache {
```
[the remark this answers is not in the document]
Comment (mxwli) [reply tools/python_api/src_cpp/include/cached_import/py_cached_import.h:8]: I didn't know which namespace to put it in so I created a new one. I'll move it to namespace kuzu.

File: tools/python_api/src_cpp/py_connection.cpp (hunk 2 above)
Comment (mxwli) [reply tools/python_api/src_cpp/py_connection.cpp]: I suppose this means I don't have to hold the gil for the last two imports?

Comment (mxwli) [reply tools/python_api/src_cpp/py_connection.cpp]: Duckdb seems to hold the GIL regardless of whether or not it was cached. The acquire statement is just found somewhere else, usually a few function calls ahead of the current function

File: tools/python_api/src_cpp/cached_import/py_cached_import.cpp [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+namespace kuzu {
+
+PythonCachedImport::~PythonCachedImport() {
+ 	py::gil_scoped_acquire acquire;
+	allObjects.clear();
+}
+
+py::handle PythonCachedImport::addToCache(py::object obj) {
+    auto ptr = obj.ptr();
+    allObjects.push_back(obj);
```
Comment (Riolku) [inline tools/python_api/src_cpp/cached_import/py_cached_import.cpp:12]: This invokes a copy.

File: tools/python_api/src_cpp/cached_import/py_cached_item.cpp [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+namespace kuzu {
+
+py::handle PythonCachedItem::operator()() {
+    assert((bool)PyGILState_Check());
+    // load if unloaded, return cached object if already loaded
+    if (loaded) {
+        return object;
+    }
+    if (parent == nullptr) {
+        object = importCache->addToCache(std::move(py::module::import(name.c_str())));
```
Comment (Riolku) [inline tools/python_api/src_cpp/cached_import/py_cached_item.cpp:16]: This also invokes a copy, doesn't it?

File: tools/python_api/src_cpp/pandas/pandas_analyzer.cpp [hunk 7]
```diff
@@ -88,8 +89,8 @@ static py::object findFirstNonNull(const py::handle& row, uint64_t numRows) {
 
 common::LogicalType PandasAnalyzer::innerAnalyze(py::object column, bool& canConvert) {
     auto numRows = py::len(column);
-    auto pandasModule = py::module::import("pandas");
-    auto pandasSeries = pandasModule.attr("core").attr("series").attr("Series");
+    auto pandasModule = importCache->pandas;
```
Comment (Riolku) [inline tools/python_api/src_cpp/pandas/pandas_analyzer.cpp:92]: Uhhhhhh this might be a copy?

**Review by Riolku — COMMENTED** [review]
Not a comprehensive list, but these are some copies I think, and they need to hold the GIL,.

File: tools/python_api/src_cpp/cached_import/py_cached_item.cpp (hunk 6 above)
Comment (mxwli) [reply tools/python_api/src_cpp/cached_import/py_cached_item.cpp:16]: There's a GIL assertion in the function

File: tools/python_api/src_cpp/pandas/pandas_analyzer.cpp (hunk 7 above)
Comment (mxwli) [reply tools/python_api/src_cpp/pandas/pandas_analyzer.cpp:92]: `pandas` isn't a py object

File: tools/python_api/src_cpp/cached_import/py_cached_import.cpp (hunk 5 above)
Comment (mxwli) [reply tools/python_api/src_cpp/cached_import/py_cached_import.cpp:12]: This is only called from functions that have the GIL. Maybe we could make the function private?

**Pull request merged**